File: skinning_test.cs
using System;
using System.Runtime.InteropServices;
using System.Diagnostics;

[StructLayout(LayoutKind.Sequential)]
public struct CalVector4 {
    public float X, Y, Z, W;

    public CalVector4 (float x, float y, float z, float w = 0.0f) {
        X = x;
        Y = y;
        Z = z;
        W = w;
    }
}

[StructLayout(LayoutKind.Sequential)]
public struct BoneTransform {
    public CalVector4 RowX, RowY, RowZ;

    public BoneTransform Scaled (float s) {
        return new BoneTransform {
            RowX = new CalVector4(RowX.X * s, RowX.Y * s, RowX.Z * s, RowX.Z * s),
            RowY = new CalVector4(RowY.X * s, RowY.Y * s, RowY.Z * s, RowY.Z * s),
            RowZ = new CalVector4(RowZ.X * s, RowZ.Y * s, RowZ.Z * s, RowZ.Z * s),
        };
    }

    public void AddScaled (ref BoneTransform toAdd, float s) {
        RowX.X += toAdd.RowX.X * s;
        RowX.Y += toAdd.RowX.Y * s;
        RowX.Z += toAdd.RowX.Z * s;
        RowX.W += toAdd.RowX.W * s;
        RowY.X += toAdd.RowY.X * s;
        RowY.Y += toAdd.RowY.Y * s;
        RowY.Z += toAdd.RowY.Z * s;
        RowY.W += toAdd.RowY.W * s;
        RowZ.X += toAdd.RowZ.X * s;
        RowZ.Y += toAdd.RowZ.Y * s;
        RowZ.Z += toAdd.RowZ.Z * s;
        RowZ.W += toAdd.RowZ.W * s;
    }

    public void TransformPoint (out CalVector4 result, ref CalVector4 point) {
        result = new CalVector4(
            (RowX.X * point.X) + (RowX.Y * point.Y) + (RowX.Z * point.Z) + RowX.W,
            (RowY.X * point.X) + (RowY.Y * point.Y) + (RowY.Z * point.Z) + RowY.W,
            (RowZ.X * point.X) + (RowZ.Y * point.Y) + (RowZ.Z * point.Z) + RowZ.W
        );
    }

    public void TransformVector (out CalVector4 result, ref CalVector4 vector) {
        result = new CalVector4(
            (RowX.X * vector.X) + (RowX.Y * vector.Y) + (RowX.Z * vector.Z),
            (RowY.X * vector.X) + (RowY.Y * vector.Y) + (RowY.Z * vector.Z),
            (RowZ.X * vector.X) + (RowZ.Y * vector.Y) + (RowZ.Z * vector.Z)
        );
    }
}

[StructLayout(LayoutKind.Sequential)]
public struct Influence {
    public int BoneId;
    public float Weight;
    public bool LastInfluenceForThisVertex;
}

[StructLayout(LayoutKind.Sequential)]
public struct Vertex {
    public CalVector4 Position, Normal;
}

public static class SkinningTest {
    public static void CalculateVerticesAndNormals (
        BoneTransform[] boneTransforms,
        int vertexCount,
        Vertex[] vertices,
        Influence[] influences,
        CalVector4[] output
    ) {
        Debug.Assert(output.Length == vertices.Length * 2);

        BoneTransform totalTransform;

        for (
            int sourceVertex = 0, sourceInfluence = 0, outputVertex = 0; 
            sourceVertex < vertices.Length; 
            sourceVertex++, sourceInfluence++, outputVertex += 2
        ) {
            var influence = influences[sourceInfluence];

            totalTransform = boneTransforms[influence.BoneId].Scaled(influence.Weight);

            while (!influence.LastInfluenceForThisVertex) {
                sourceInfluence += 1;
                influence = influences[sourceInfluence];

                totalTransform.AddScaled(
                    ref boneTransforms[influence.BoneId], influence.Weight
                );
            }

            totalTransform.TransformPoint(
                out output[outputVertex], ref vertices[sourceVertex].Position
            );
            totalTransform.TransformVector(
                out output[outputVertex + 1], ref vertices[sourceVertex].Normal
            );
        }
    }

    public static void Main () {
        const int N = 10000;
        const long SecondInTicks = 10000000;

        var vertices = new Vertex[N];
        var influences = new Influence[N];

        for (int i = 0; i < N; i++) {
            vertices[i] = new Vertex {
                Position = new CalVector4(1, 2, 3),
                Normal = new CalVector4(0, 0, 1)
            };

            influences[i] = new Influence {
                BoneId = 0,
                Weight = 1.0f,
                LastInfluenceForThisVertex = true
            };
        }

        var boneTransforms = new BoneTransform[] {
            new BoneTransform()
        };

        var output = new CalVector4[N * 2];

        for (int i = 0; i < 100; i++) {
            long verticesSkinned = 0;
            long started = DateTime.UtcNow.Ticks;

            while (DateTime.UtcNow.Ticks < (started + SecondInTicks)) {
                CalculateVerticesAndNormals(
                    boneTransforms, N, vertices, influences, output
                );

                verticesSkinned += N;
            }

            long elapsed = DateTime.UtcNow.Ticks - started;

            Console.WriteLine(
                "Skinned vertices per second: {0}",
                (verticesSkinned * SecondInTicks) / elapsed
            );
        }
    }
}

File: skinning_test.cpp
// compiled in cygwin with:
// g++ -Wall -O2 -o skinning_test skinning_test.cpp

#include <vector>
#include <set>
#include <map>
#include <assert.h>
#include <math.h>
#include <xmmintrin.h>
#include <string.h>
#include <stdio.h>
#include <time.h>

#define CAL3D_ALIGN_HEAD(N)
#define CAL3D_ALIGN_TAIL(N)

struct CalBase4 {
  float x, y, z, w;

  void set(float _x, float _y, float _z, float _w) {
    x = _x;
    y = _y;
    z = _z;
    w = _w;
  }
}
CAL3D_ALIGN_TAIL(16);

struct CalVector4 : CalBase4 {
  CalVector4() {
    x = 0.0f;
    y = 0.0f;
    z = 0.0f;
    w = 0.0f;
  }

  CalVector4(float x, float y, float z, float w = 0.0f) {
    this->x = x;
    this->y = y;
    this->z = z;
    this->w = w;
  }

  void setAsVector(float x, float y, float z) {
    this->x = x;
    this->y = y;
    this->z = z;
    w = 0.0f;
  }
};

struct CalPoint4 : CalBase4 {
  CalPoint4() {
    x = 0.0f;
    y = 0.0f;
    z = 0.0f;
    w = 1.0f;
  }

  CalPoint4(float x, float y, float z, float w = 1.0f) {
    this->x = x;
    this->y = y;
    this->z = z;
    this->w = w;
  }

  void setAsPoint(float x, float y, float z) {
    x = this->x;
    y = this->y;
    z = this->z;
    w = 1.0f;
  }
};

// 3x3 transform matrix plus a translation 3-vector (stored in the w components
// of the rows.  This struct needs to be 16-byte aligned for SSE.
struct BoneTransform {
  CalVector4 rowx;
  CalVector4 rowy;
  CalVector4 rowz;
};

struct Influence
{
  Influence() {
    boneId = -1;
    weight = 0.0f;
    lastInfluenceForThisVertex = 0;
  }
  
  Influence(unsigned b, float w, bool last) {
    boneId = b;
    weight = w;
    lastInfluenceForThisVertex = last ? 1 : 0;
  }

  unsigned boneId;
  float weight;
  unsigned lastInfluenceForThisVertex;
};

CAL3D_ALIGN_HEAD(16)
struct Vertex
{
  CalPoint4 position;
  CalVector4 normal;
}
CAL3D_ALIGN_TAIL(16);


void calculateVerticesAndNormals_SSE_intrinsics(
  const BoneTransform* boneTransforms,
  int vertexCount,
  const Vertex* vertices,
  const Influence* influences,
  CalVector4* output_vertex
) {

  __m128 rowx;
  __m128 rowy;
  __m128 rowz;
  __m128 weight;

  // calculate all submesh vertices
  while (vertexCount--) {
    weight = _mm_load_ss(&influences->weight);
    weight = _mm_shuffle_ps(weight, weight, _MM_SHUFFLE(0, 0, 0, 0));

    const BoneTransform& bt = boneTransforms[influences->boneId];

    rowx = _mm_mul_ps(_mm_load_ps((const float*)&bt.rowx), weight);
    rowy = _mm_mul_ps(_mm_load_ps((const float*)&bt.rowy), weight);
    rowz = _mm_mul_ps(_mm_load_ps((const float*)&bt.rowz), weight);

    while (!influences++->lastInfluenceForThisVertex) {
      weight = _mm_load_ss(&influences->weight);
      weight = _mm_shuffle_ps(weight, weight, _MM_SHUFFLE(0, 0, 0, 0));

      const BoneTransform& bt = boneTransforms[influences->boneId];

      rowx = _mm_add_ps(rowx, _mm_mul_ps(_mm_load_ps((const float*)&bt.rowx), weight));
      rowy = _mm_add_ps(rowy, _mm_mul_ps(_mm_load_ps((const float*)&bt.rowy), weight));
      rowz = _mm_add_ps(rowz, _mm_mul_ps(_mm_load_ps((const float*)&bt.rowz), weight));
    }

    {
      // transform position
      const __m128 position = _mm_load_ps((const float*)&vertices->position);
      
      const __m128 mulx = _mm_mul_ps(position, rowx);
      const __m128 muly = _mm_mul_ps(position, rowy);
      const __m128 mulz = _mm_mul_ps(position, rowz);

      const __m128 copylo = _mm_unpacklo_ps(mulx, muly);
      const __m128 copyhi = _mm_unpackhi_ps(mulx, muly);
      const __m128 sum1 = _mm_add_ps(copylo, copyhi);

      const __m128 lhps = _mm_movelh_ps(mulz, sum1);
      const __m128 hlps = _mm_movehl_ps(sum1, mulz);
      const __m128 sum2 = _mm_add_ps(lhps, hlps);

      _mm_storeh_pi((__m64*)output_vertex, sum2);

      __m128 sum3 = _mm_shuffle_ps(sum2, sum2, _MM_SHUFFLE(1, 1, 1, 1));
      sum3 = _mm_add_ss(sum2, sum3);
      _mm_store_ss(&output_vertex->z, sum3);
    }

    // transform normal
    {
      const __m128 normal = _mm_load_ps((const float*)&vertices->normal);
      
      const __m128 mulx = _mm_mul_ps(normal, rowx);
      const __m128 muly = _mm_mul_ps(normal, rowy);
      const __m128 mulz = _mm_mul_ps(normal, rowz);

      const __m128 copylo = _mm_unpacklo_ps(mulx, muly);
      const __m128 copyhi = _mm_unpackhi_ps(mulx, muly);
      const __m128 sum1 = _mm_add_ps(copylo, copyhi);

      const __m128 lhps = _mm_movelh_ps(mulz, sum1);
      const __m128 hlps = _mm_movehl_ps(sum1, mulz);
      const __m128 sum2 = _mm_add_ps(lhps, hlps);

      _mm_storeh_pi((__m64*)&output_vertex[1], sum2);

      __m128 sum3 = _mm_shuffle_ps(sum2, sum2, _MM_SHUFFLE(1, 1, 1, 1));
      sum3 = _mm_add_ss(sum2, sum3);
      _mm_store_ss(&output_vertex[1].z, sum3);
    }

    ++vertices;
    output_vertex += 2;
  }
}

int main() {
  const int N = 10000;

  Vertex v[N];
  Influence i[N];
  for (int k = 0; k < N; ++k) {
    v[k].position.setAsPoint(1.0f, 2.0f, 3.0f);
    v[k].normal.setAsVector(0.0f, 0.0f, 1.0f);
    i[k].boneId = 0;
    i[k].weight = 1.0f;
    i[k].lastInfluenceForThisVertex = true;
  }

  BoneTransform bt;
  memset(&bt, 0, sizeof(bt));

  CAL3D_ALIGN_HEAD(16) CalVector4 output[N * 2] CAL3D_ALIGN_TAIL(16);

  long long vertices_skinned = 0;
  
  clock_t start = clock();
  while (clock() < start + CLOCKS_PER_SEC) {
    calculateVerticesAndNormals_SSE_intrinsics(&bt, N, v, i, output);
    vertices_skinned += N;
  }
  clock_t elapsed = clock() - start;

  float sum = 0;
  for (unsigned i = 0; i < N * 2; ++i) {
    sum += (output[i].x + output[i].y + output[i].z + output[i].w);
  }

  printf("Skinned vertices per second: %d, blah=%f\n", (int)(vertices_skinned * CLOCKS_PER_SEC / elapsed), sum);
}
